### PR TITLE
Do not use github.sha in pull_request and workflow_run

### PR DESCRIPTION
**Pull request by EeroPaukkonen**
Issues in the previous version
* The `version` was determined by `github.sha`. This can point to an undesired location (e.g. workflow_run will always point to HEAD of the default branch) and there currently is no way to override this.
* Another issue is that we do checkout without ref inside the action so even if you checkout some other version we are not uploading that. Checkout action will just checkout to [github.ref](https://github.com/actions/checkout/blob/main/src/input-helper.ts#L61-L72).

Fixes: https://linear.app/odacom/issue/CP-1112/upload-velo-artifacts-using-githubsha-as-version-break-things

This PR adds 
* Ability to override sha.
* defaults to `pull_request` and `workflow_run` (So that users do not need to know these)
* ~~version input that can override sha versioning (not really yet supported elsewhere so we could leave it out if we want)~~ (Removed because it needs to be much more complicated and does not fit this PR)

This PR should only affect existing workflows that use `pull_request` and `workflow_run`. Most likely only in a positive way.
* [pull_request](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) will not reference the merge branch commits but the actual commit end result. 
* [workflow_run](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) Would have been really broken if not running in the default branch (main). But in the default branch we will be only incrementally better (pointing to the right commit not just HEAD). 

### Use ref instead of sha
* We could input ref instead of sha
* After checking out to ref we can get sha from `git rev-parse HEAD`
* ➕ More flexible
* ➕ More aligned with checkout action.
* ➖ Should we also have ref in deploy-action?
* ➖ Is this more confusing to understand?

### No checkout inside upload-velo-artifacts
* We could remove checkout from `upload-velo-artifacts` and require users of the action to checkout. 
* And then just grab the sha from `git rev-parse`. This is essentially what [velo-action](https://github.com/kolonialno/velo-action/blob/main/velo_action/version.py#L15) does.
* ➕ This would make checkout more explicit. Currently, people might not realize we are doing a checkout inside the action which could mess with the following steps. 
* ➖ This would mean that we cannot provide easy default values for ref/sha.
* ➖ This would mean that we need to add checkout to quite a few existing `upload-velo-artifacts@main`.

### Original version-fun (this idea does not work)
* https://github.com/kolonialno/upload-velo-artifacts/pull/4/files was intended to rely on checkouts done outside of the action. 
* We did not notice that `upload-velo-artifacts` was doing checkout inside the action which broke things.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -14,11 +14,31 @@ inputs:
...
+  version:
+    description: "By default version is same as sha."
+    required: false
 runs:
   using: "composite"
   steps:
+    - name: Get sha
+      shell: bash
+      run: |
+        echo "TARGET_SHA=${{ inputs.sha || github.event.pull_request.head.sha || github.event.workflow_run.head_sha || github.sha }}" >> $GITHUB_ENV
```
Comment (thor) [inline action.yml:37]: hey, that's pretty cool -- learnt something new, didn't know you could OR them like in React.

**Comment by thor** [comment]
Looking this morning.

**Comment by EeroPaukkonen** [comment]
Tested here https://github.com/kolonialno/velodrome-testing/pull/11

**Comment by EeroPaukkonen** [comment]
@andersliland @thor Opinions on this PR?

File: README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@
   A GitHub Action to upload Velo deployment artifacts
 </p>
 
-This GitHub actions is part of the Velo deploy system at Oda.
+Upload-velo-artifacts is part of the Velo deploy system at Oda. It first checkouts to `sha` and then uploads `deploy_folder`.
```
Comment (thor) [inline README.md:10]:
```suggestion
`upload-velo-artifacts` is part of the Velo deploy system at Oda. It first performs a checkout of `sha` and then uploads `deploy_folder`.
```
nitpick: language/formatting

File: action.yml [hunk 3]
```diff
@@ -14,11 +14,28 @@ inputs:
     description: "Name of the Google cloud storage bucket to upload the artifacts to"
     required: false
     default: ".deploy"
+  sha:
+    description: |
+      Sha used to checkout artifacts.
```
Comment (thor) [inline action.yml:19]:
```suggestion
      SHA used to checkout the repository.
```
nitpick/suggestion

File: action.yml [hunk 4]
```diff
@@ -14,11 +14,28 @@ inputs:
     description: "Name of the Google cloud storage bucket to upload the artifacts to"
     required: false
     default: ".deploy"
+  sha:
+    description: |
+      Sha used to checkout artifacts. 
+      If sha is not provided even trigger is matched to sha following way.
```
Comment (thor) [inline action.yml:20]: unsure suggestion: ?
```suggestion
      If `sha` is not provided, then the triggering event is matched to the SHA the following way.
```

**Review by thor — APPROVED** [review]
Only insignificant nitpicks/unsure suggestions. LGTM.

**Pull request merged**